module/module.go

```diff
@@ -0,0 +1,422 @@
+	"go.viam.com/rdk/protoutils"
+	"go.viam.com/rdk/registry"
+	"go.viam.com/rdk/resource"
+	"go.viam.com/rdk/robot/client"
+	"go.viam.com/rdk/subtype"
+)
+
+// HandlerMap is the format for api->model pairs that the module will service.
+// Ex: rdk:component:motor -> [acme:marine:thruster, acme:marine:outboard].
+type HandlerMap map[resource.RPCSubtype][]resource.Model
+
+// ToProto converts the HandlerMap to a protobuf representation.
+func (h HandlerMap) ToProto() *pb.HandlerMap {
+	pMap := &pb.HandlerMap{}
+	for s, models := range h {
+		subtype := &robotpb.ResourceRPCSubtype{
+			Subtype: protoutils.ResourceNameToProto(resource.Name{
+				Subtype: s.Subtype,
+				Name:    "",
+			}),
+			ProtoService: s.SvcName,
+		}
+
+		handler := &pb.HandlerDefinition{Subtype: subtype}
+		for _, m := range models {
+			handler.Models = append(handler.Models, m.String())
+		}
+		pMap.Handlers = append(pMap.Handlers, handler)
+	}
+	return pMap
+}
+
+// NewHandlerMapFromProto converts protobuf to HandlerMap.
+func NewHandlerMapFromProto(ctx context.Context, pMap *pb.HandlerMap, conn *grpc.ClientConn) (HandlerMap, error) {
+	hMap := make(HandlerMap)
+	refClient := grpcreflect.NewClient(ctx, reflectpb.NewServerReflectionClient(conn))
+	defer refClient.Reset()
+	reflSource := grpcurl.DescriptorSourceFromServer(ctx, refClient)
+
+	var errs error
+	for _, h := range pMap.GetHandlers() {
+		api := protoutils.ResourceNameFromProto(h.Subtype.Subtype).Subtype
+
+		symDesc, err := reflSource.FindSymbol(h.Subtype.ProtoService)
+		if err != nil {
+			errs = multierr.Combine(errs, err)
+			continue
+		}
+		svcDesc, ok := symDesc.(*desc.ServiceDescriptor)
+		if !ok {
+			return nil, errors.Errorf("expected descriptor to be service descriptor but got %T", symDesc)
+		}
+		subtype := &resource.RPCSubtype{
+			Subtype: api,
+			Desc:    svcDesc,
+		}
+		for _, m := range h.Models {
+			model, err := resource.NewModelFromString(m)
+			if err != nil {
+				return nil, err
+			}
+			hMap[*subtype] = append(hMap[*subtype], model)
+		}
+	}
+	return hMap, errs
+}
+
+// Module represents an external resource module that services components/services.
+type Module struct {
+	parent                  *client.RobotClient
+	grpcServer              *grpc.Server
+	logger                  *zap.SugaredLogger
+	mu                      sync.Mutex
+	ready                   bool
+	addr                    string
+	parentAddr              string
+	activeBackgroundWorkers sync.WaitGroup
+	handlers                HandlerMap
+	services                map[resource.Subtype]subtype.Service
+	pb.UnimplementedModuleServiceServer
+}
+
+// NewModule returns the basic module framework/structure.
+func NewModule(address string, logger *zap.SugaredLogger) *Module {
+	m := &Module{
+		logger:     logger,
+		addr:       address,
+		grpcServer: grpc.NewServer(),
+		ready:      true,
+		handlers:   make(HandlerMap),
+		services:   make(map[resource.Subtype]subtype.Service),
+	}
+	m.buildHandlerMap()
```

Comment:
This builds a handlermap dynamically from things in the registry. In other languages/SDKs, it may make more sense to just let the module author define the map statically.